File: 0254-factor-combinations/0254-factor-combinations.py
class Solution:
    def __init__(self):
        self.result = []
    
    def dfs(self, curr, n, candidates, currList):
        # print(n, candidates[curr], candidates, currList)
        if n % candidates[curr] == 0:
            currList.append(candidates[curr])
            currList.append(int(n / candidates[curr]))
            self.result.append(currList.copy())
            currList.pop()
            newCandidates = [x for x in range(candidates[curr], floor(sqrt(n / candidates[curr]))+1)]
            # print(newCandidates)
            for i in range(len(newCandidates)):
                self.dfs(i, int(n / candidates[curr]), newCandidates, currList)
            currList.pop()

    def getFactors(self, n: int) -> List[List[int]]:
        # if n == 32:
        #     return [[2,16],[4,8],[2,2,8],[2,4,4],[2,2,2,4],[2,2,2,2,2]]
        candidates = [x for x in range(2, floor(sqrt(n))+1)]
        for i in range(len(candidates)):
            self.dfs(i, n, candidates, [])
        return self.result

File: 0254-factor-combinations/0254-factor-combinations.java
class Solution {
    private List<List<Integer>> result;

    private void breakDown(LinkedList<Integer> currList) {
        if (currList.size() > 1) {
            this.result.add(new LinkedList<>(currList));
        }
        int lastElement = currList.removeLast();
        for (int i = currList.size() == 0 ? 2 : currList.peekLast(); i <= lastElement / i; i++) {
            if (lastElement % i == 0) {
                currList.addLast(i);
                currList.addLast(lastElement / i);
                this.breakDown(currList);
                currList.removeLast();
                currList.removeLast();
            }
        }
        currList.addLast(lastElement);
    }

    public List<List<Integer>> getFactors(int n) {
        this.result = new LinkedList<>();
        this.breakDown(new LinkedList<Integer>(Arrays.asList(n)));  
        return this.result;  
    }
}